### ISSUE TITLE
How do you define "unsupported request"?

### ISSUE THREAD

**Issue by Relequestual** [json-api/json-api#1548]
CRUD updating section specifies...

> A server MUST return 403 Forbidden in response to an unsupported request to update a resource or relationship.

https://jsonapi.org/format/1.1/#crud-updating-relationship-responses-403

How do you define "unsupported request" here?

In the application I work with, routes that are not defined result in a 404. The entity could be known, but the route not defined and so a 404 is returned.

For us: 
- if a user doesn't have permissions to see something, they get a 404.
- If a user tries to update a known entity that isn't allowed to be updated, because we haven't defined the route, it's a 404.
- If a user tries to update something they are allowed to update, but they have something wrong in the payload, it results in 400 or 409.

In what situation do you expect a 403 to be returned?

**Comment by dgeb** [comment]
I think the response sections are overdue to be revisited. I believe that the highest order instruction is buried in `Other Responses`:

>  A server MUST prepare responses, and a client MUST interpret responses, in accordance with HTTP semantics.

I think we just confuse the matter by placing that last. And I believe all the responses you listed are valid and compliant with HTTP. 

> In what situation do you expect a 403 to be returned?

I might return a 403 if someone can view a resource but not modify it due to permissions. If someone can't view a resource due to permissions, a 404 is perfectly reasonable.

@gabesullice and I have recently discussed reworking all the response sections for 1.1.

**Comment by Relequestual** [comment]
I hear what you're saying, but the order in which requirements are laid out wouldn't make a difference here.

Sure, what I've said we do is in accordance with HTTP semantics, but other clauses which say "MUST", are a requirement, regardless of other clauses.


> 
>     In what situation do you expect a 403 to be returned?
> 
> I might return a 403 if someone can view a resource but not modify it due to permissions.

That makes sense. We have an interesting situation where the API user can't see anything outside of what they can edit with their permissions when using the API. As I've said, there are other reasons why someone might not be able to edit something, like you simply aren't allowed to edit that entity using the API at all. In that situation it IS technically a permissions issue, but a 403 woudln't make sense, as the route simply doesn't exist (404).

Strictly, as I read the API, edit routes should always be provided, and permission denied (for everyone), returning 403, give our situation. (I accept this was likely not the intent.)

I apologies if I'm maybe more critical than others might be... I'm a spec author myself, so I nitpick these sorts of details.

**Comment by dgeb** [comment]
No need to apologize! Let's dig into this ...

> Strictly, as I read the API, edit routes should always be provided, and permission denied (for everyone), returning 403, give our situation. (I accept this was likely not the intent.)

To edit a resource, the spec says:

>  A resource can be updated by sending a PATCH request to the URL that represents the resource.

From the spec's perspective, the "URL that represents the resource" is provided by a `self` link on a resource object appearing in a prior response document (according to the spec, `self` "identifies the resource represented by the resource object.")

If such a URL is provided to a user, and permission to edit is denied at that URL, you MUST return a 403 according to the spec.

However, if such a URL is never exposed to a user, then I would interpret the spec to allow returning 404 to be a perfectly valid response. And I think that covers the scenario you describe here:

> We have an interesting situation where the API user can't see anything outside of what they can edit with their permissions when using the API. As I've said, there are other reasons why someone might not be able to edit something, like you simply aren't allowed to edit that entity using the API at all. In that situation it IS technically a permissions issue, but a 403 woudln't make sense, as the route simply doesn't exist (404).

**Comment by Relequestual** [comment]
Well...
We have an entity that the user is allowed to read, but not update.
We define the GET route, but no other methods on the path are defined.

**Comment by freddrake** [comment]
That's an interesting case, and I know I have cases like that as well.

In my application, there are routes that have the corresponding PATCH defined, if anyone is allowed to edit, but they'll return the 403 if the authenticated user is not allowed to edit.  Other routes represent read-only resources, and those will return a 405 (Method Not Allowed), as an accident of implementation (with DELETE behaving similarly).

Strictly speaking, these are non-compliant with the specification.

**Comment by Relequestual** [comment]
Right, I think there are a number of parts of the spec where I feel there could be tighter defined requirements, or more aligned with HTTP semantics.

When you have a configurator generator API (You configure what you want, and the routes are all generated for you), then you can do a lot automatically, but building those, for complex systems, is not only extremely difficult (I've done it a few times), but horrifically costly, in comparison to defining the routes manually (for a small number of entities). If you have a large number of entities, then the cost may be worthwhile.

**Comment by freddrake** [comment]
Depends on the frameworks you're using.  I believe in my case (Python + Flask), compliance on this aspect would be easy to achieve using an appropriate base class for endpoints.  It's just not a high priority since the current response achieves the required result (deny disallowed updates).

The required 403 seems well-aligned to me; it's just not the only way to achieve the desired result.  Constraining it to a 403 and not a 405 works well, and potentially makes client implementation (ever so slightly) easier.

**Comment by dgeb** [comment]
> We have an entity that the user is allowed to read, but not update.
> We define the GET route, but no other methods on the path are defined.

A 403 seems most appropriate to me here too, since the user can read the resource from that same URL. I do think it adds clarity above and beyond a 404 without compromising security.

**Comment by Relequestual** [comment]
> > We have an entity that the user is allowed to read, but not update.
> > We define the GET route, but no other methods on the path are defined.
> 
> A 403 seems most appropriate to me here too, since the user can read the resource from that same URL. I do think it adds clarity above and beyond a 404 without compromising security.

I don't disagree. It just seems odd to define a route that always responds with a 403.

**Comment by gabesullice** [comment]
> We have an entity that the user is allowed to read, but not update. We define the GET route, but no other methods on the path are defined.

A [`405 Method Not Allowed`](https://httpstatuses.com/405) would definitely be the most appropriate response here, if you didn't know anything else about the spec. A `404` is definitely inappropriate because the server _can_ locate the resource, by definition (it can `GET` it).

> Right, I think there are a number of parts of the spec where I feel there could be tighter defined requirements, or more aligned with HTTP semantics.

Agreed. This is the discussion that @dgeb was mentioning in his [earlier comment](https://github.com/json-api/json-api/issues/1548#issuecomment-802822738).

---

I think I can explain the confusion (which should not be interpreted as me saying the spec is fine as-is).

Consider the following request:

```http
PATCH /article/1

{
  "data": {
    "relationships": {
      "author": {
        "data": {
          "type": "user",
          "id": 1
        }
      }
    }
  }
}
```

If the client is not allowed to update the `author` relationship on article 1, as shown above, the spec says the appropriate response is a `403 Forbidden`. That's because the `PATCH` method might be allowed for other fields, but not the `author` field specifically. An [error object](https://jsonapi.org/format/#error-objects) ought to convey this distinction with a JSON pointer to the `author` field. 

Similarly, consider the following request:

```http
PATCH /article/1/relationships/author

{
  "data": {
    "type": "user",
    "id": 1
  }
}
```

The spec is implicitly treating the request above as interchangeable with the request before it (notice the URL). This is because relationship endpoints are meant to be a kind of _window_ or scoped view of the parent resource's relationship field. IOW, relationship resources represent relationship fields. We can infer that from the way that the `self` link is defined for relationships objects:

> `self`: a link for the relationship itself (a “relationship link”). This link allows the client to **directly manipulate** the relationship. 

IOW, both request strategies for updating a resource's relationship are meant to be interchangeable since a client might prefer one over the other. For example, a `POST` to a to-many relationship resource is non-destructive when simultaneous requests are a concern. OTOH, a `PATCH` to the same field via the parent resource object's URL would be unsafe.

I think the `403` rule in the OP is attempting to enforce a symmetry between the resource field and the relationship resource that represents it. Without that background knowledge, I agree that it seems to be breaking HTTP semantics though.

---

**tl;dr**: the confusion stems from whether you think of the relationship resource as its own resource (yes, technically it is) or as a "window" into the base resource's relationship field.

The problem here is not the definition of what an "unsupported request" is, it's that JSON:API's philosophy of standalone relationship resources is not well explained via its normative language.

**Comment by freddrake** [comment]
@Relequestual: I agree that seems a little odd.  And in the generic case (where a 404 / 405 is acceptable), there's no need.

But how odd it is depends very much on what it means to register a route, which gets back to the frameworks being used.

**Comment by Relequestual** [comment]
I don't think I have any confusion over the relationships requirements laid out in the spec.

The `responses` section appears to relate to the WHOLE "updating" section, but I do note the anchor for the 403 is `#crud-updating-relationship-responses-403`. Was it intended that the 403 response rule only applies to updating relationships and not resources in general?

If it is intended that the 403 response only relates to relationships, then "unsupported request" makes sense, however in the case of updates in general, I still find I'm unsure what it means.



---

In our specific situation, this one object type doesn't provide any `links` object in its GET response, so one could argue that a dynamic client would have no basis on which to make any further requests against the resource returned 🤷‍♂️.

More an aside, but the relationships thing for us is difficult. There are no parts of our system exposed to the API that allows the creating, updating, or deleting of a relationship by itself. We have strict referential integrity and relational requirements.

**Comment by lode** [comment]
I have a similar situation where we don't want to offer a `GET /<type>/<id>`.

We offer a `GET /parent/<id>/relationship` to get a list of items, and we do offer a `PATCH|DELETE /<type>/<id>`, so the resources in the list have a self link pointing to `/<type>/<id>` to explain how the PATCH|DELETE can be reached. However, the GET on that endpoint will give a 406/404/403.

It feels a little weird though to pass a self link which can't be used for a GET.
